### ISSUE TITLE
Double escaped SSH commands

### ISSUE THREAD

**Issue by johnbillion** [wp-cli/wp-cli#4650]
When running a command over SSH (eg. via the `--ssh` flag or an `@`-alias), the contents of the command flag appears to be double escaped.

Example:

    wp --ssh=foo plugin list --debug

Results in:

    Running SSH command: ssh -q 'foo' -t 'wp '\''plugin'\'' '\''list'\'' '\''--debug'\'''

I would expect it to be:

    Running SSH command: ssh -q 'foo' -t 'wp plugin list --debug'

The double escaping appears to come from two calls to `escapeshellarg`, [first inside `WP_CLI\Runner::run_ssh_command()`](https://github.com/wp-cli/wp-cli/blob/b264acab3e9f117f8c39926ea82f6be4e6207521/php/WP_CLI/Runner.php#L445) and secondly within [the corresponding condition in `WP_CLI\Runner::generate_ssh_command()`](https://github.com/wp-cli/wp-cli/blob/b264acab3e9f117f8c39926ea82f6be4e6207521/php/WP_CLI/Runner.php#L525) (and conditions above that for Docker, Vagrant, etc).

Is there a reason for the double escaping that I'm missing?

**Comment by gitlost** [comment]
Well I think it's just simpler to program, to escape once for the local shell and once for the remote. Is it causing an issue?

**Comment by johnbillion** [comment]
It doesn't cause a problem, but it's also not ideal. A program should be aware of the correct places to apply appropriate escaping, not apply escaping just in case.

**Comment by danielbachhuber** [comment]
> Is there a reason for the double escaping that I'm missing?

Yes: I wrote bad code :) It's worth fixing.

**Comment by marksabbath** [comment]
I've been looking into this and it does seem that it's scaping the arguments and commands after they are receiving the `'`, causing that weirdness.

IMHO running the `escapeshellarg` would be required for each argument, but **before** the argument parse (or after, but removing the `'` at the begging and at the end of the strings).

Does this approach sounds reasonable for you guys?

**Comment by schlessera** [comment]
> after they are receiving the `'`

@marksabbath I'm not sure what you mean by that? I don't think we're directly adding quotes anywhere, they are only added by `escapeshellarg()`. Or am I missing something?

**Comment by marksabbath** [comment]
@schlessera you're correct. It's `escapeshellarg()` that's adding the `'` for each argument. I'll keep diggging into this to find a good solution.
